datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxRecommendationService.java

```diff
@@ -0,0 +1,194 @@
+package com.sequenceiq.datalake.service.sdx;
+
+import static com.sequenceiq.cloudbreak.common.exception.NotFoundException.notFound;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import javax.inject.Inject;
+
+import org.apache.commons.lang3.ObjectUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+import com.sequenceiq.cloudbreak.api.endpoint.v4.stacks.request.StackV4Request;
+import com.sequenceiq.cloudbreak.api.endpoint.v4.stacks.request.instancegroup.InstanceGroupV4Request;
+import com.sequenceiq.cloudbreak.common.exception.BadRequestException;
+import com.sequenceiq.cloudbreak.common.json.JsonUtil;
+import com.sequenceiq.cloudbreak.common.mappable.CloudPlatform;
+import com.sequenceiq.common.api.type.CdpResourceType;
+import com.sequenceiq.datalake.configuration.CDPConfigService;
+import com.sequenceiq.datalake.converter.VmTypeConverter;
+import com.sequenceiq.datalake.entity.SdxCluster;
+import com.sequenceiq.datalake.service.EnvironmentClientService;
+import com.sequenceiq.environment.api.v1.environment.model.response.DetailedEnvironmentResponse;
+import com.sequenceiq.environment.api.v1.platformresource.model.PlatformVmtypesResponse;
+import com.sequenceiq.sdx.api.model.SdxClusterShape;
+import com.sequenceiq.sdx.api.model.SdxDefaultTemplateResponse;
+import com.sequenceiq.sdx.api.model.SdxRecommendationResponse;
+import com.sequenceiq.sdx.api.model.VmTypeMetaJson;
+import com.sequenceiq.sdx.api.model.VmTypeResponse;
+
+@Service
+public class SdxRecommendationService {
```

Comment:
logic of this class' public methods should be placed in a try catch probably, some of them called from controller. I see several chained get calls, if an NPE or similar occurs, we should transform it into user friendly error